File: non-linear/creamery/label.py
"""Helpers for building formulation problems from CPG-style ingredient labels."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Mapping, Sequence
import math

from .analysis import NutritionFacts
from .constants import K_F_WATER, MW_SUCR, MIX_DENSITY_KG_PER_L, PINT_L
from .model import (
    Constraint,
    FormulationGoals,
    FormulationProblem,
    FormulationSolution,
)


@dataclass(frozen=True)
class LabelGroup:
    """A label line; keys are ingredient variables aggregated for ordering."""

    name: str
    keys: Sequence[str]
    fraction_bounds: Mapping[str, tuple[float, float]] = field(default_factory=dict)
    enforce_internal_order: bool = False


def goals_from_label(
    facts: NutritionFacts,
    pint_mass_g: float,
    *,
    serve_temp_C: float = -12.0,
    draw_temp_C: float = -5.0,
    shear_rate_s: float = 50.0,
) -> FormulationGoals:
    """Convert a nutrition panel and pint mass into target mix goals."""

    mass_kg = pint_mass_g / 1000.0
    servings = pint_mass_g / facts.serving_size_g
    fat = facts.total_fat_g * servings / 1000.0
    carbs = facts.total_carbs_g * servings / 1000.0
    sugars = facts.total_sugars_g * servings / 1000.0
    protein = facts.protein_g * servings / 1000.0
    ash = (facts.sodium_mg * servings / 1000.0) * (58.44 / 23.0) / 1000.0  # Na→NaCl kg
    solids = fat + carbs + protein + ash
    water = max(1e-6, mass_kg - solids)

    sucrose_moles = sugars * 1000.0 / MW_SUCR
    fp_est = -K_F_WATER * sucrose_moles / water
    finished_density = mass_kg / PINT_L
    overrun_guess = max(0.0, min(1.5, MIX_DENSITY_KG_PER_L / finished_density - 1.0))

    return FormulationGoals(
        batch_mass=100.0,
        fat_pct=fat / mass_kg,
        solids_pct=solids / mass_kg,
        sweetness_pct=sugars / mass_kg,
        freezing_point_C=float(fp_est),
        overrun=overrun_guess,
        serve_temperature_C=serve_temp_C,
        draw_temperature_C=draw_temp_C,
        shear_rate_s=shear_rate_s,
    )


def apply_group_bounds(problem: FormulationProblem, groups: Sequence[LabelGroup]) -> None:
    """Add within-group fraction bounds (e.g., cream fat share, liquid sugar solids)."""

    for group in groups:
        group_total = problem.sum_of(group.keys)
        for key, (lo, hi) in group.fraction_bounds.items():
            # If the group has a single member, fraction bounds are meaningless (always 1.0).
            if len(group.keys) == 1 and group.keys[0] == key:
                continue
            problem.add_constraint(problem.expr(key) - hi * group_total, -math.inf, 0.0, f"{group.name}:{key}:hi")
            problem.add_constraint(problem.expr(key) - lo * group_total, 0.0, math.inf, f"{group.name}:{key}:lo")


def apply_label_order(problem: FormulationProblem, groups: Sequence[LabelGroup], epsilon: float = 1e-3) -> None:
    """Enforce descending label order for groups and within-group members."""

    ordered = [g.keys for g in groups]
    problem.add_ordering(ordered, epsilon=epsilon)

    for group in groups:
        if group.enforce_internal_order and len(group.keys) > 1:
            within = [[key] for key in group.keys]
            problem.add_ordering(within, epsilon=epsilon)


__all__ = [
    "LabelGroup",
    "goals_from_label",
    "apply_group_bounds",
    "apply_label_order",
]

File: label.go
package creamery

import "math"

// LabelGroup bundles ingredients that appear together on a consumer label.
type LabelGroup struct {
	Name                 string
	Keys                 []IngredientID
	FractionBounds       map[IngredientID]Interval
	EnforceInternalOrder bool
}

// LabelGoals captures macro targets inferred from a Nutrition Facts panel.
type LabelGoals struct {
	BatchMassKG      float64
	FatPct           float64
	SolidsPct        float64
	SweetnessPct     float64
	FreezingPointC   float64
	Overrun          float64
	ServeTemperature float64
	DrawTemperature  float64
	ShearRate        float64
	OverrunCap       *float64
}

// GoalsFromLabel converts Nutrition Facts + pint mass into formulation goals.
func GoalsFromLabel(facts NutritionFacts, pintMassGrams float64, serveTemp, drawTemp, shearRate float64) LabelGoals {
	if serveTemp == 0 {
		serveTemp = defaultServeTempC
	}
	if drawTemp == 0 {
		drawTemp = defaultDrawTempC
	}
	if shearRate == 0 {
		shearRate = defaultShearRate
	}

	massKg := pintMassGrams / 1000.0
	servings := pintMassGrams / facts.ServingSizeGrams
	fat := facts.TotalFatGrams * servings / 1000.0
	carbs := facts.TotalCarbGrams * servings / 1000.0
	sugars := facts.TotalSugarsGrams * servings / 1000.0
	protein := facts.ProteinGrams * servings / 1000.0
	ash := (facts.SodiumMg * servings / 1000.0) * (58.44 / 23.0) / 1000.0
	solids := fat + carbs + protein + ash
	water := math.Max(1e-6, massKg-solids)

	sucroseMoles := sugars * 1000.0 / mwSucrose
	fpEst := -kfWater * sucroseMoles / water
	finishedDensity := massKg / pintLiters
	overrunGuess := math.Max(0.0, math.Min(1.5, mixDensityKgPerL/finishedDensity-1.0))

	return LabelGoals{
		BatchMassKG:      100.0,
		FatPct:           fat / massKg,
		SolidsPct:        solids / massKg,
		SweetnessPct:     sugars / massKg,
		FreezingPointC:   fpEst,
		Overrun:          overrunGuess,
		ServeTemperature: serveTemp,
		DrawTemperature:  drawTemp,
		ShearRate:        shearRate,
	}
}

// ApplyGroupBounds enforces within-group fraction bounds.
func ApplyGroupBounds(p *Problem, groups []LabelGroup) {
	for _, group := range groups {
		if len(group.Keys) == 0 {
			continue
		}
		for keyID, bounds := range group.FractionBounds {
			keyName, ok := p.nameForID(keyID)
			if !ok {
				continue
			}
			if len(group.Keys) == 1 && group.Keys[0] == keyID {
				continue
			}
			// key <= hi * group_total
			if bounds.Hi < math.Inf(1) {
				coeffs := make(map[string]float64, len(group.Keys)+1)
				coeffs[keyName] = 1
				for _, memberID := range group.Keys {
					if memberName, ok := p.nameForID(memberID); ok {
						coeffs[memberName] -= bounds.Hi
					}
				}
				p.AddConstraint(coeffs, math.Inf(-1), 0, group.Name+":"+keyName+":hi")
			}
			// key >= lo * group_total
			if bounds.Lo > 0 {
				coeffs := make(map[string]float64, len(group.Keys)+1)
				coeffs[keyName] = 1
				for _, memberID := range group.Keys {
					if memberName, ok := p.nameForID(memberID); ok {
						coeffs[memberName] -= bounds.Lo
					}
				}
				p.AddConstraint(coeffs, 0, math.Inf(1), group.Name+":"+keyName+":lo")
			}
		}
	}
}

// ApplyLabelOrder enforces descending weight order of label groups (and optionally members).
func ApplyLabelOrder(p *Problem, groups []LabelGroup, epsilon float64) {
	if epsilon <= 0 {
		epsilon = 1e-3
	}
	for i := 0; i < len(groups)-1; i++ {
		earlier := groups[i]
		later := groups[i+1]
		if len(earlier.Keys) == 0 || len(later.Keys) == 0 {
			continue
		}
		coeffs := make(map[string]float64, len(earlier.Keys)+len(later.Keys))
		for _, key := range later.Keys {
			if name, ok := p.nameForID(key); ok {
				coeffs[name] = 1
			}
		}
		for _, key := range earlier.Keys {
			if name, ok := p.nameForID(key); ok {
				coeffs[name] -= 1
			}
		}
		p.AddConstraint(coeffs, math.Inf(-1), -epsilon, "label_order")
	}

	for _, group := range groups {
		if !group.EnforceInternalOrder {
			continue
		}
		for i := 0; i < len(group.Keys)-1; i++ {
			a := group.Keys[i]
			b := group.Keys[i+1]
			aName, aOK := p.nameForID(a)
			bName, bOK := p.nameForID(b)
			if !aOK || !bOK {
				continue
			}
			coeffs := map[string]float64{
				bName: 1,
				aName: -1,
			}
			p.AddConstraint(coeffs, math.Inf(-1), 0, group.Name+":internal")
		}
	}
}
